tests/test_url_shortening.py

```diff
@@ -0,0 +1,206 @@
+        db.session.remove()
+        db.drop_all()
+        self.app_context.pop()
+
+    def api_email_auth_headers(self, email, password):
+        """Setup an authorization header with an email and password."""
+        return {
+            'Authorization': 'Basic ' + b64encode(
+                (email + ':' + password).encode('utf-8')).decode('utf-8'),
+            'Accept': 'application/json',
+            'Content-Type': 'application/json'
+        }
+
+    def api_token_auth_headers(self, token):
+        """Setup an authorization header with a token."""
+        return {
+            'Authorization': 'Basic ' + b64encode(
+                (token + ':' + '').encode('utf-8')).decode('utf-8'),
+            'Accept': 'application/json',
+            'Content-Type': 'application/json'
+        }
+
+    def test_anonymous_user_shorten_via_api(self):
+        """Test that anonymous users cannot shorten urls via the API."""
+        response = self.client.post(url_for('api.shorten_url'),
+                                    data=self.post_data)
+        json_response = json.loads(response.data.decode('utf-8'))
+        msg = json_response['message']
+        self.assertEqual(msg, 'Permission required! You are not allowed'
+                         ' access to this resource.')
+        self.assertEqual(403, response.status_code)
+
+    def test_registered_user_shorten_via_api(self):
+        """Test that registered users can shorten urls via the API."""
+        self.user.save()
+        email_header = self.api_email_auth_headers('ichiato@yahoo.com',
+                                                   'password')
+        token_response = self.client.get(url_for('api.get_token'),
+                                         headers=email_header)
+        json_token_response = json.loads(token_response.data.decode('utf-8'))
+        token = json_token_response['token']
+        token_header = self.api_token_auth_headers(token)
+        shorten_response_with_email_auth = self.client.post(
+            url_for('api.shorten_url'), headers=email_header,
+            data=self.post_data)
+        shorten_response_with_token_auth = self.client.post(
+                            url_for('api.shorten_url'), headers=token_header,
+                            data=self.post_data)
+        json_shorten_response_with_email = json.loads(
+            shorten_response_with_email_auth.data.decode('utf-8'))
+        email_msg = json_shorten_response_with_email['message']
+        json_shorten_response_with_token = json.loads(
+            shorten_response_with_token_auth.data.decode('utf-8'))
+        self.assertEqual(email_msg, 'Permission required! You are not allowed'
+                         ' access to this resource.')
+        self.assertIn('short_url', json_shorten_response_with_token)
+        self.assertEqual(403, shorten_response_with_email_auth.status_code)
+        self.assertEqual(201, shorten_response_with_token_auth.status_code)
```

Comment:
While it should be possible for registered users to access the api endpoints with tokens, their email and password combination should be rejected. That was what I was trying to test. I would be breaking this method into two and then get rid of the long names to give it clarity. Thanks though